### PR TITLE
#1491 remove global usage

### LINKED ISSUE #1491: Remove use of global object
## Is your feature request related to a problem? Please describe.

`AMC` calculation uses the `global` object in `Settings`

## Describe the solution you'd like

Not 100% sure of the solution, but certainly removing `global` is the goal. The tissue should look into implementation

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1491 

## Change summary

Main goal of this PR was to remove the usage of the `global` object in `MobileAppSettings`. This meant changing usage logic. Ended up extracting the consumption logic to a method outside  - reasoning
- Having this logic extracted is easier if (when) we need to change usage logic. Will be far easier to create a new algorithm and avoid regressions (i.e. could use an `IF customisation: usage1() else usage2()`). "Pseudo dependency injection" - i'd argue a lot of the business logic should be extracted in such a way
- The logic was a little convoluted. Calling helpers from `Item` -> `ItemBatch` -> `TransactionBatch` was a bit hard to follow, and they've never been reused. Is a lot easier to digest in a small sequential method, in my opinion

Additional changes that sneaked in:
- Saving all fields that are in `[store]custom_data` which effect mobile as their own settings keys [This is an initial step in hopefully, a better way to organise/manage this as it is inevitable this is going to be used more(?)]
- Instead of refresh the `global` object, 'refresh' the settings keys in `PostSyncProcessor`. I wouldv'e liked to use the processor more as it is intended, but we don't have `Store` in schema, so unable to queue up functions for that data type.
- When refreshing the `global` object, also accounts for *removing* keys, which was not being done before
- Also added a sync action for completion of sync, (I shouldn't have done this in this PR, sorry) such that the module reducer can subscribe and update the interface if modules are changed. This was not being done before, you had to restart the app)

## Testing

Steps to reproduce or otherwise test the changes of this PR:

*Note: Please let me know if further details are needed if this is ever tested*

- [ ] Usage calculation is the 'same as before'!
- [ ] using `monthsLeadTime` custom data is synced (Can use realm browser)
- [ ] using `monthlyConsumptionEnforceLookBackPeriod` custom data is synced (Can use realm browser)
- [ ] using `monthlyConsumptionLookBackPeriod` custom data is synced (Can use realm browser)
- [ ] using `usesDispensaryModule` custom data is synced (Can use realm browser)
- [ ] using `usesDispensaryModule` custom data is synced (Can use realm browser)

### Related areas to think about

- Need to write docs for all this custom data stuff
- Really wanted to rename the monthly consumption settings but would have to write redux persistance migration code etc. Might do in a new issue, but might not worry about it. 
- Found #1519 during this
- In a way this is also starting to address #880 - I would like to see a `DataType` field added to `Settings`.. so it would be nicer to manage different data types

### CODE REVIEW

File: src/mSupplyMobileApp.js [hunk 1]
```diff
@@ -151,6 +152,7 @@ class MSupplyMobileAppContainer extends React.Component {
     } else {
       this.postSyncProcessor.processRecordQueue();
     }
+    dispatch(syncCompleteTransaction());
```
Comment (joshxg) [inline src/mSupplyMobileApp.js]: This naming doesn't really make sense. Want a way to signal 'after post processing sync', but don't want to pass dispatch to post sync processor

File: src/pages/MenuPage.js [hunk 2]
```diff
@@ -56,7 +56,7 @@ const Menu = ({
 
   const MenuButton = useCallback(
     props => <Button style={menuButton} textStyle={buttonText} {...props} />,
-    []
+    [usingDispensary, usingModules]
```
Comment (joshxg) [inline src/pages/MenuPage.js]: This will be messy and not good if there's ever 4+ modules..

File: src/settings/index.js [hunk 3]
```diff
@@ -18,6 +18,19 @@ export const SETTINGS_KEYS = {
   THIS_STORE_NAME_ID: 'ThisStoreNameId',
   THIS_STORE_TAGS: 'ThisStoreTags',
   THIS_STORE_CUSTOM_DATA: 'ThisStoreCustomData',
+  CONSUMPTION_LOOKBACK_PERIOD: 'monthlyConsumptionLookBackPeriod',
+  CONSUMPTION_ENFORCE_LOOKBACK: 'monthlyConsumptionEnforceLookBackPeriod',
+  MONTHS_LEAD_TIME: 'monthsLeadTime',
+  VACCINE_MODULE: 'usesVaccineModule',
+  DISPENSARY_MODULE: 'usesDispensaryModule',
```
Comment (joshxg) [inline src/settings/index.js]: `uses` `using` and no `prefix` is confusing

File: src/sync/PostSyncProcessor.js [hunk 4]
```diff
@@ -83,6 +86,39 @@ export class PostSyncProcessor {
...
+   * After syncing, recheck and save all of these fields.
+   */
+  processCustomData = () => {
+    // This stores custom data, stored as a JSON stringified object
+    const customData = this.settings.get(SETTINGS_KEYS.THIS_STORE_CUSTOM_DATA);
+
+    try {
+      // If there is no custom data, use an empty object and clear all
+      // customizable settings (set to an empty string)
+      const parsedCustomData = JSON.parse(customData || '{}');
```
Comment (joshxg) [inline src/sync/PostSyncProcessor.js]: this is pretty ugly

File: src/utilities/dailyUsage.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+/**
+ * Returns if this store is customized such that usage should force a lookback period,
+ * rather than using an items added date.
+ */
+const getAMCEnforcementLookback = () => {
+  const amcEnforcementString = UIDatabase.getSetting('monthlyConsumptionEnforceLookBackPeriod');
+  return amcEnforcementString === 'true';
+};
```
Comment (joshxg) [inline src/utilities/dailyUsage.js]: this combined with the `checkModule` helper in the `ModuleReducer`.. lots of spread out helpers

File: src/utilities/dailyUsage.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  const numberOfUsageDays = millisecondsToDays(usagePeriod);
+
+  const usage = UIDatabase.objects('TransactionBatch')
+    .filtered('itemBatch.item.id == $0', itemId)
+    .filtered('transaction.type == $0', 'customer_invoice')
+    .filtered('transaction.confirmDate >= $0 && transaction.confirmDate <= $1', startDate, dateNow)
+    .reduce((sum, { totalQuantity }) => sum + totalQuantity, 0);
+
+  return usage / numberOfUsageDays || 1;
+};
```
Comment (joshxg) [inline src/utilities/dailyUsage.js]: Wondering if it would be better to split this into two (three) functions - `dailyUsage` which would be a splitter which would call either `normalUsage` or `usageWithLookbackEnforcement` (or `usageWithItemAddedLogic`, `usageJustOverTheLastXDays`..) .. maybe not for now

File: src/database/DataTypes/ItemBatch.js [hunk 7]
```diff
@@ -80,25 +78,7 @@ export class ItemBatch extends Realm.Object {
...
-      startDate,
-      endDate
-    );
-
-    return getTotal(transactionBatches, 'usage');
-  }
-
-  /**
-   * Add a transaction batch to be associated with this batch.
+   * Ad˚d a transaction batch to be associated with this batch.
```
Comment (mark-prins) [inline src/database/DataTypes/ItemBatch.js]:
```suggestion
   * Add a transaction batch to be associated with this batch.
```

File: src/utilities/dailyUsage.js (hunk 6 above)
Comment (mark-prins) [reply src/utilities/dailyUsage.js]: yeah, not for now. it's ok as is
